book/course/theory.md

```diff
@@ -0,0 +1,195 @@
+* Downloading from package repositories.
+* installation and removal of packages.
+* Upgrading/downgrading package versions.
+* Installation and removal of dependancies.
+
+The most commonly used package managers on ARC3/ARC4 are anaconda and pip.
+
+* [Anaconda](https://arcdocs.leeds.ac.uk/software/compilers/anaconda.html) is an environment and package manager, primarily for Python and R.
+* [pip](https://en.wikipedia.org/wiki/Pip_(package_manager)) is a package manager for Python.
+
+## Build tools
+
+Build tools is a generic term for any programs and tools used to build software. Building software can include steps such as configuration, compiling, testing, packaging.
+
+Common buildtools :
+
+* Autotools - is a set of build tools to help make portal Unix-like software.
+* CMake - is a cross platform build automation tool, to support the creation of configurations for build tools.
+* Make - the traditional Unix/Linux build tool.
+
+### Compilers
+
+A compiler is a program which translates source code written in a programming language into target code. This is typically into machine code to create an executable program.
+
+ARC3 and ARC4 have the [Intel](https://arcdocs.leeds.ac.uk/software/compilers/intel.html) compiler set as default, with the [GNU](https://arcdocs.leeds.ac.uk/software/compilers/gnu.html) and [LLVM](https://arcdocs.leeds.ac.uk/software/compilers/llvm.html) compilers being popular alternatives available on the module system.
+
+#### Interpreters
+
+An interpreter reads and executes source code directly. They generally runs code slower than a compiled executable as it has to compile code every time the code is run, but there are optimisations such as bytecode compilation and caching that can vastly improve performance.
+
+Python is a popular interpreted language and the most common interpreter is the default, [CPython](https://en.wikipedia.org/wiki/CPython).
+
+Some languages are capable of being compiled or run through an interpreter.
+
+### Build Configuration
+
+Tools like CMake can be used to create different platform specific for different build tools configurations.
+
+Configuration files usually contain locations of the source code, output locations, of existing components (libraries, assets such as help documentation, data).
+
+They can contain options for conditional builds such as for debug and release builds and for additional features to take advantage of optional components such as input/output format/interfaces or hardware - such as GPUs.
+
+### Packaging
+
+```{Note}
+Users do not have sufficent permissions to install system packages on our HPC systems.
+Containers provide a mechanism for users to install system packages in a user managed environment.
+```
+
+Linux and package distribtions usually provide tools to automate many steps of the process of creating software packages from source; this will include steps to compile source code, package the files required and test the package is installable/removable.
```

Comment:
```suggestion
Linux and package distributions usually provide tools to automate many steps of the process of creating software packages from source; this will include steps to compile source code, package the files required and test the package is installable/removable.
```